src/solve.jl

```diff
@@ -100,62 +113,78 @@ function _solve(cp::ControlledPlant,
     FT = Flowpipe{NT, RT, Vector{RT}}
-    out = Vector{FT}(undef, NSAMPLES)
+    flowpipes = Vector{FT}()
+    controls = Vector()
+
+    # waiting list
+    waiting_list = Vector{ControlFlowpipe{FT}}()
 
-    for i in 1:NSAMPLES
-        if i == 1
+    # iterate
+    first = true
+    while true
+        if first
             X = nothing
             X₀ = project(Q₀, st_vars)
+            k = 1
+            first = false
+        elseif isempty(waiting_list)
+            break
         else
-            X = sol(ti)
-            X₀ = _project_oa(X, st_vars, ti;
+            prev_part = pop!(waiting_list)
+            k = prev_part.k + 1
+            t = tend(prev_part.F)
+            X = prev_part.F(t)
+            X₀ = _project_oa(X, st_vars, t;
                              remove_zero_generators=remove_zero_generators) |> set
         end
-        P₀ = m == 0 ? X₀ : X₀ × W₀
-
-        # get new control inputs from the controller
-        Us = Vector{splitter.output_type}()
-        for X₀ in split(splitter, X₀)
-            X0aux = apply(preprocessing, X₀)
-            U₀ = forward_network(solver, network, X0aux)
-            U₀ = apply(normalization, U₀)
-            push!(Us, U₀)
+        t0 = tvec[k]
+        t1 = tvec[k+1]
+        F, U = _solve_one(X, X₀, W₀, S, st_vars, t0, t1, cpost, nnet_forward,
+                            rec_method)
+        push!(flowpipes, F)
+        push!(controls, U)
+        if k < length(tvec) - 1
+            push!(waiting_list, ControlFlowpipe(F, k))
         end
-        U₀ = merge(splitter, UnionSetArray(Us))
+    end
 
-        # simplify the control input for intervals
-        if dim(U₀) == 1
-            U₀ = overapproximate(U₀, Interval)
-        end
+    ext = Dict{Symbol, Any}(:controls=>controls)
+    return MixedFlowpipe(flowpipes, ext)
+end
 
-        Q₀ = _reconstruct(rec_method, P₀, U₀, X, ti)
+function _solve_one(X, X₀, W₀, S, st_vars, t0, t1, cpost, nnet_forward, rec_method)
+    # add nondeterministic inputs (if any)
+    P₀ = isnothing(W₀) ? X₀ : X₀ × W₀
 
-        Ti = i < NSAMPLES ? (ti + sampling_time + Δti) : tend(time_span)
+    # get new control inputs from the controller
+    U₀ = nnet_forward(X₀)
 
-        controls[i] = U₀
-        dt = ti .. Ti
-        sol = post(cpost, IVP(S, Q₀), dt)
-        out[i] = sol
+    # combine states with new control inputs
+    Q₀ = _reconstruct(rec_method, P₀, U₀, X, t0)
 
-        ti = tend(sol)
-        Δti = Ti - ti  # difference of exact and actual control time
-        @assert LazySets.isapproxzero(Δti) "the flowpipe duration differs " *
-            "from the requested duration by $Δti time units (stopped at $ti)"
-    end
+    dt = t0 .. t1
+    sol = post(cpost, IVP(S, Q₀), dt)
 
-    ext = Dict{Symbol, Any}(:controls=>controls)
-    return MixedFlowpipe(out, ext)
+    t1′ = tend(sol)
+    Δt = t1 - t1′  # difference of exact and actual control time
+    @assert LazySets.isapproxzero(Δt) "the flowpipe duration differs " *
```

Comment:
not directly related to this PR, but i think that here we should have the more relaxed check `t1 \in tspan(sol)` -- because that's all that is needed to evaluate it. (sometimes it happens that check fails because of rounding though, but there is special code in RA to handle it.. so i wouldn't mind if we remove the assertion / check about time domains altogether.. if it doesn't fail here, it will fail in the evaluation in time of the solution)

Reply:
I suggest we do this in another change.